File: src/driver/io/PCA9557.cpp
#include "PCA9557.hpp"
#include <iostream>

#ifndef BIT
#define BIT(nr)                 (1UL << (nr))
#endif
#define AT581X_STATUS_Calibration_Enable    (3)             /* 1 --Calibration enable; 0 --Calibration disable */
#define AT581X_STATUS_CRC_FLAG              (4)
#define AT581X_STATUS_BUSY_INDICATION       (7)
using namespace std;
using namespace be;

namespace be::driver::io {
    DEFINE_NCLASS_META(PCA9557, I2CDevice)

    std::vector<JSCFunctionListEntry> PCA9557::methods = {
            JS_CFUNC_DEF("read", 0, PCA9557::read),
            JS_CFUNC_DEF("write", 0, PCA9557::write),
            JS_CFUNC_DEF("readMode", 0, PCA9557::readMode),
            JS_CFUNC_DEF("writeMode", 0, PCA9557::writeMode),
            JS_CFUNC_DEF("getPinMode", 0, PCA9557::getPinMode),
            JS_CFUNC_DEF("setPinMode", 0, PCA9557::setPinMode),
    };

    PCA9557::PCA9557(JSContext *ctx, JSValue _jsobj)
            : I2CDevice(ctx, build(ctx, _jsobj))
    {
        regVal_Mode = 0xFF ;
        regVal_Value = 0xFF ;
    }

    JSValue PCA9557::constructor(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
        auto obj = new PCA9557(ctx);
        return obj->jsobj;
    }
    
    int PCA9557::setup() {
        if(!i2c) {
            return -1 ;
        }
        if(!readMode()) {
            return  -2 ;
        }
        return 0 ;
    }

    bool PCA9557::read(uint8_t pin, uint8_t & val) {
        if(!i2c || !addr) {
            return false ;
        }
        if(pin>7) {
            return false ;
        }
        if(!i2c->read<uint8_t,uint8_t>(addr, 0, regVal_Value)){
            return false ;
        }
        val = (regVal_Value >> pin) & 1 ;
        return true ;
    }

    bool PCA9557::write(uint8_t pin, uint8_t val) {
        if(!i2c || !addr) {
            return false ;
        }
        if(pin>7) {
            return false ;
        }
        if(val) {
            regVal_Value|= 1<<pin ;
        } else {
            regVal_Value&= ~(1<<pin) ;
        }
        i2c->write(addr, 1, regVal_Value) ;
        return true ;
    }
    
    uint8_t PCA9557::getMode() const {
        return regVal_Mode ;
    }

    bool PCA9557::readMode() {
        if(!i2c || !addr) {
            return false ;
        }
        if(!i2c->read<uint8_t,uint8_t>(addr, 3, regVal_Mode)){
            return false ;
        }
        return true ;
    }

    bool PCA9557::writeMode(uint8_t mode) {
        if(!i2c || !addr) {
            return false ;
        }
        regVal_Mode = mode ;
        i2c->write(addr, 3, regVal_Mode) ;
        return true ;
    }
    bool PCA9557::getPinMode(uint8_t pin, Mode & mode) {
        if(!i2c || !addr) {
            return false ;
        }
        if(pin>7) {
            return false ;
        }
        if(!i2c->read<uint8_t,uint8_t>(addr, 3, regVal_Mode)){
            return false ;
        }
        mode = (regVal_Mode & (1<<pin))? Mode::input: Mode::output ;
        return true ;
    }
    bool PCA9557::setPinMode(uint8_t pin, Mode mode) {
        if(!i2c || !addr) {
            return false ;
        }
        if(pin>7) {
            return false ;
        }
        if(mode==Mode::input) {
            regVal_Mode|= (1<<pin) ;
        } else {
            regVal_Mode&= ~(1<<pin) ;
        }
        return i2c->write(addr, 3, regVal_Mode) ;
    }

    JSValue PCA9557::read(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
        return JS_UNDEFINED ;
    }
    JSValue PCA9557::write(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
        return JS_UNDEFINED ;
    }
    JSValue PCA9557::readMode(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
        THIS_NCLASS(PCA9557, thisobj)
        if(!thisobj->readMode()) {
            JSTHROW("%s.%s() failed", "PCA9557", "readMode")
        }
        return JS_NewUint32(ctx, thisobj->getMode()) ;
    }
    JSValue PCA9557::writeMode(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
        CHECK_ARGC(1)
        THIS_NCLASS(PCA9557, thisobj)
        ARGV_TO_UINT8(0, mode)
        if(!thisobj->writeMode(mode)) {
            JSTHROW("%s.%s() failed", "PCA9557", "writeMode")
        }
        return JS_UNDEFINED ;
    }
    JSValue PCA9557::getPinMode(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
        return JS_UNDEFINED ;
    }
    JSValue PCA9557::setPinMode(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
        return JS_UNDEFINED ;
    }
}

File: src/js/driver/io/PCA9557.js
import * as driver from 'driver'
import { exportValue } from "loader"

const mapModeValue = {
    output: 0 ,
    input: 1
}
// mode: 1=输入, 0=输出
class PCA9557 extends driver.I2CDevice {
    mode = 0
    output = 0
    alias={}
    setup(opt) {
        if(!opt.addr) {
            opt.addr = 0x1c
        }
        super.setup(opt)
        if(opt.mode!=undefined) {
            this.setMode(opt.mode)
        } else {
            this.getMode(true)
        }
        if(opt.alias) {
            for(let p=0;p<8;p++){
                if(opt.alias[p]) {
                    this.alias[opt.alias[p]] = p
                }
            }
        }
    }
    read(pin) {
        let [val] = this.readReg(0)
        if(this.alias[pin]!=undefined) {
            pin = this.alias[pin]
        }
        return (val & (1<<pin))? 1: 0
    }
    write(pin,level) {
        if(this.alias[pin]!=undefined) {
            pin = this.alias[pin]
        }
        let output = (this.output & ~(1 << pin)) | (level << pin)
        this.writeReg(1, output)
        this.output = output
    }
    readAllPins() {
        let [val] = this.readReg(0)
        return val
    }
    writeAllPins(val) {
        this.writeReg(1, val)
        this.output = val
    }
    getMode(sync) {
        if(sync) {
            [this.mode] = this.readReg(3)
        }
        return this.mode
    }
    setMode(mode) {
        this.writeReg(3,mode)
        this.mode = mode
    }
    getPinMode(pin,sync) {
        if(sync) {
            this.getMode(true)
        }
        if(this.alias[pin]!=undefined) {
            pin = this.alias[pin]
        }
        console.log("pin")
        return (this.mode & (1<<pin))? "input": "ouput"
    }
    setPinMode(pin,dir,sync) {
        if(this.alias[pin]!=undefined) {
            pin = this.alias[pin]
        }
        if(mapModeValue[dir]!=undefined) {
            dir = mapModeValue[dir]
        }
        if(dir!=1 && dir!=0) {
            throw new Exception("invalid pin dir, must be 1(input), 0(output)")
        }
        if(sync) {
            this.getMode(true)
        }
        if(dir) {
            this.setMode(this.mode | (1<<pin))
        } else {
            this.setMode(mode & (~(1<<pin)))
        }
    }
    printMode() {
        for(let i=0;i<8;i++) {
            console.log(`pin[${i}]: ${this.getPinMode(i)}`)
        }
    }
}

exportValue(driver, "PCA9557", PCA9557)
